// Generated from "trycatch/TryCatch.java"
@file:OptIn(ExperimentalObjCName::class, ExperimentalObjCRefinement::class)
@file:Suppress(
 "ALWAYS_NULL",
 "PARAMETER_NAME_CHANGED_ON_OVERRIDE",
 "REPEATED_BOUND",
 "SENSELESS_COMPARISON",
 "UNCHECKED_CAST",
 "UNNECESSARY_LATEINIT",
 "UNNECESSARY_NOT_NULL_ASSERTION",
 "UNREACHABLE_CODE",
 "UNUSED_PARAMETER",
 "UNUSED_VARIABLE",
 "USELESS_CAST",
 "VARIABLE_IN_SINGLETON_WITHOUT_THREAD_LOCAL",
 "VARIABLE_WITH_REDUNDANT_INITIALIZER")

package trycatch

import javaemul.lang.*
import java.lang.AutoCloseable
import java.lang.ClassCastException
import javaemul.internal.Exceptions
import kotlin.Exception
import kotlin.NullPointerException
import kotlin.OptIn
import kotlin.RuntimeException
import kotlin.Suppress
import kotlin.Throwable
import kotlin.assert
import kotlin.experimental.ExperimentalObjCName
import kotlin.experimental.ExperimentalObjCRefinement
import kotlin.jvm.JvmField
import kotlin.native.HiddenFromObjC
import kotlin.native.ObjCName

@ObjCName("J2ktTrycatchTryCatch", exact = true)
open class TryCatch {
 @ObjCName("testMultiCatch")
 open fun testMultiCatch() {
  try {
   throw ClassCastException()
  } catch (e: NullPointerException) {
   throw e!!
  } catch (e: kotlin.ClassCastException) {
   throw e!!
  } catch (r: RuntimeException) {
   var r_1: RuntimeException? = r
   r_1 = null
  }
 }

 @ObjCName("testEmptyThrowableCatch")
 open fun testEmptyThrowableCatch() {
  try {
   throw ClassCastException()
  } catch (e: Throwable) {}
  try {
   throw ClassCastException()
  } catch (e_1: Exception) {} catch (e_2: Throwable) {}
 }

 @ObjCName("testEmptyThrowableRethrow")
 open fun testEmptyThrowableRethrow() {
  try {
   throw ClassCastException()
  } catch (e: Throwable) {
   throw e!!
  }
 }

 @ObjCName("testFinally")
 open fun testFinally() {
  try {
   assert(true)
  } finally {}
 }

 @ObjCName("testTryWithResource")
 open fun testTryWithResource() {
  try {
   var ___primaryExc: Throwable? = null
   var thing: TryCatch.ClosableThing? = null
   var thing2: TryCatch.ClosableThing? = null
   try {
    thing = TryCatch.ClosableThing()
    thing2 = TryCatch.ClosableThing()
    throw java.lang.Exception()
   } catch (___exceptionFromTry: Throwable) {
    ___primaryExc = ___exceptionFromTry
    throw ___exceptionFromTry!!
   } finally {
    ___primaryExc = Exceptions.safeClose(thing2, ___primaryExc)
    ___primaryExc = Exceptions.safeClose(thing, ___primaryExc)
    if (___primaryExc != null) throw ___primaryExc!!
   }
  } catch (e: Exception) {}
 }

 @ObjCName("testTryWithResourceJava9")
 open fun testTryWithResourceJava9() {
  val thing: TryCatch.ClosableThing? = TryCatch.ClosableThing()
  val thing2: TryCatch.ClosableThing? = TryCatch.ClosableThing()
  try {
   var ___primaryExc: Throwable? = null
   var ___resource: TryCatch.ClosableThing? = null
   var ___resource_1: TryCatch.ClosableThing? = null
   try {
    ___resource = thing
    ___resource_1 = thing2
    throw java.lang.Exception()
   } catch (___exceptionFromTry: Throwable) {
    ___primaryExc = ___exceptionFromTry
    throw ___exceptionFromTry!!
   } finally {
    ___primaryExc = Exceptions.safeClose(___resource_1, ___primaryExc)
    ___primaryExc = Exceptions.safeClose(___resource, ___primaryExc)
    if (___primaryExc != null) throw ___primaryExc!!
   }
  } catch (e: Exception) {}
 }

 @ObjCName("testTryWithResouceOnStaticField")
 open fun testTryWithResouceOnStaticField() {
  try {
   var ___primaryExc: Throwable? = null
   var ___resource: TryCatch.ClosableThing? = null
   try {
    ___resource = TryCatch.closableThing_private_2
    throw java.lang.Exception()
   } catch (___exceptionFromTry: Throwable) {
    ___primaryExc = ___exceptionFromTry
    throw ___exceptionFromTry!!
   } finally {
    ___primaryExc = Exceptions.safeClose(___resource, ___primaryExc)
    if (___primaryExc != null) throw ___primaryExc!!
   }
  } catch (e: Exception) {}
 }

 @ObjCName("testNestedTryCatch")
 open fun testNestedTryCatch() {
  try {
   throw java.lang.Exception()
  } catch (ae: Exception) {
   try {
    throw java.lang.Exception()
   } catch (ie: Exception) {}
  }
 }

 @ObjCName("testThrowGenerics")
 open fun testThrowGenerics() {
  throw this.getT_private_2<Exception?>(java.lang.Exception())!!
 }

 @HiddenFromObjC
 fun <T> getT_private_2(t: T): T {
  return t
 }

 @ObjCName("testThrowBoundGenerics")
 open fun testThrowBoundGenerics() {
  throw this.getThrowable_private_2<Throwable?>()!!
 }

 @HiddenFromObjC
 fun <T: Throwable?> getThrowable_private_2(): T {
  return null as T
 }

 @ObjCName("J2ktTrycatchTryCatchCompanion", exact = true)
 companion object {
  @JvmField
  @HiddenFromObjC
  val closableThing_private_2: TryCatch.ClosableThing? = TryCatch.ClosableThing()
 }

 open class ClosableThing: AutoCloseable {
  open override fun close() {}
 }
}
